### ISSUE TITLE
Pytorch implementation of meta updating

### ISSUE THREAD

**Issue by JaminFong** [dragen1860/MAML-Pytorch#9]
Hi dragen,
   I have read your code. It's a good job.
   When I tried to implement the maml, I used `parameters_origin = model.state_dict()` and `model.load_state_dict(parameters_origin)` to regain the weights after inner update. And then I carried out meta update using the summed loss. But this seemed not to work, and the acc waved around 20.
   I don't know what's going wrong and have u reproduced the result on miniimagenet?
Best!

**Comment by dragen1860** [comment]
hi,
`model.state_dict()` only works for standard `nn.module` class. However, my `model` class is not a standard class, hence you cant save weights by `model.state_dict()`. You can manually save the weights tensor by pickle.

**Comment by JaminFong** [comment]
I know what you mean. But I do this as follows:
`		weights_origin = copy.deepcopy(self.net.parameters())
		for i in range(batchsz): # batchsz==self.meta_batchsz
			pred_q = self.net(query_x[i], bns=None, training=training)
			pred_q = F.softmax(pred_q, dim=1).argmax(dim=1)
			correct = torch.eq(pred_q, query_y[i]).sum().item()
			corrects[0] = corrects[0] + correct

			# 1. run the i-th task and compute loss for k=0
			self.inner_optim.zero_grad()
			pred = self.net(support_x[i])
			loss = F.cross_entropy(pred, support_y[i])
			loss.backward()
			nn.utils.clip_grad_norm_(self.net.parameters(), 5)
			self.inner_optim.step()

			pred_q = self.net(query_x[i], bns=None, training=training)
			loss_q = F.cross_entropy(pred_q, query_y[i])

			pred_q = F.softmax(pred_q, dim=1).argmax(dim=1)
			correct = torch.eq(pred_q, query_y[i]).sum().item()
			corrects[1] = corrects[1] + correct

			for k in range(1, self.K):
				# 1. run the i-th task and compute loss for k=1~K-1
				self.inner_optim.zero_grad()
				pred = self.net(support_x[i], bns=None, training=training)
				loss = F.cross_entropy(pred, support_y[i])
				loss.backward()
				nn.utils.clip_grad_norm_(self.net.parameters(), 5)
				self.inner_optim.step()

				pred_q = self.net(query_x[i], bns=None, training=training)
				loss_q = F.cross_entropy(pred_q, query_y[i])
				pred_q = F.softmax(pred_q, dim=1).argmax(dim=1)
				correct = torch.eq(pred_q, query_y[i]).sum().item() # convert to numpy
				corrects[k+1] = corrects[k+1] + correct

			# 4. record last step's loss for task i
			losses_q.append(loss_q)
			for p, p_o in zip(self.net.parameters(), weights_origin):
				p.data.copy_(p_o)

		# end of all tasks
		# sum over all losses across all tasks
		loss_q = torch.stack(losses_q).sum(0)
`
I load the original weights before every task, but update the weights by optimizer rather than using `fast weights`. But the acc waved around 20. I don't know what the matter is.

**Comment by JaminFong** [comment]
I know what you mean. But I do this as follows:
```
		weights_origin = copy.deepcopy(self.net.parameters())
		for i in range(batchsz): # batchsz==self.meta_batchsz
			pred_q = self.net(query_x[i], bns=None, training=training)
			pred_q = F.softmax(pred_q, dim=1).argmax(dim=1)
			correct = torch.eq(pred_q, query_y[i]).sum().item()
			corrects[0] = corrects[0] + correct

			# 1. run the i-th task and compute loss for k=0
			self.inner_optim.zero_grad()
			pred = self.net(support_x[i])
			loss = F.cross_entropy(pred, support_y[i])
			loss.backward()
			nn.utils.clip_grad_norm_(self.net.parameters(), 5)
			self.inner_optim.step()

			pred_q = self.net(query_x[i], bns=None, training=training)
			loss_q = F.cross_entropy(pred_q, query_y[i])

			pred_q = F.softmax(pred_q, dim=1).argmax(dim=1)
			correct = torch.eq(pred_q, query_y[i]).sum().item()
			corrects[1] = corrects[1] + correct

			for k in range(1, self.K):
				# 1. run the i-th task and compute loss for k=1~K-1
				self.inner_optim.zero_grad()
				pred = self.net(support_x[i], bns=None, training=training)
				loss = F.cross_entropy(pred, support_y[i])
				loss.backward()
				nn.utils.clip_grad_norm_(self.net.parameters(), 5)
				self.inner_optim.step()

				pred_q = self.net(query_x[i], bns=None, training=training)
				loss_q = F.cross_entropy(pred_q, query_y[i])
				pred_q = F.softmax(pred_q, dim=1).argmax(dim=1)
				correct = torch.eq(pred_q, query_y[i]).sum().item() # convert to numpy
				corrects[k+1] = corrects[k+1] + correct

			# 4. record last step's loss for task i
			losses_q.append(loss_q)
			for p, p_o in zip(self.net.parameters(), weights_origin):
				p.data.copy_(p_o)

		# end of all tasks
		# sum over all losses across all tasks
		loss_q = torch.stack(losses_q).sum(0)
```
I load the original weights before every task, but update the weights by optimizer rather than using `fast weight`s. But the acc waved around 20. I don't know what the matter is.

**Comment by JaminFong** [comment]
@flexibility2 Indeed, I think every meta-batch update is based on the original parameters as the paper illustrated.
![ci7ozignrhg n uy 7rwjmu](https://user-images.githubusercontent.com/32559548/46858766-a1a46580-ce3e-11e8-95ac-c76fc3b30d5d.png)

**Comment by flexibility2** [comment]
@JaminFong 
ok, and I think you are right~ Could you share the complete code so that we can discuss with that.
In addition, do you know what "K" stands for in the line 5 of Algorithm? I can't  understand completely, and I hope I can get your explaination.

**Comment by ArnoutDevos** [comment]
@flexibility2 `K` is the amount of "shots" you sample for task T_i (line 5), which are grouped in the set `D`. I think that for every task T_i  gradient evaluation (line 6) the whole set `D` of task T_i is used.
